api/applications.go

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package api
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/google/uuid"
+	"github.com/hashicorp/go-multierror"
+	abstractions "github.com/microsoft/kiota-abstractions-go"
+	msgraphsdkgo "github.com/microsoftgraph/msgraph-sdk-go"
+	auth "github.com/microsoftgraph/msgraph-sdk-go-core/authentication"
+	"github.com/microsoftgraph/msgraph-sdk-go/applications"
+	"github.com/microsoftgraph/msgraph-sdk-go/models"
+)
+
+type ApplicationsClient interface {
```

Comment:
Great thoughts!

I'm curious about the idea of defining our own types — I had a similar thought initially, but was having a hard time figuring out how to get our constructed types to match up with the complicated interfaces (such as `models.Applicationable`) that Microsoft API calls return. Did you have ideas on how we might define the types? Would definitely make our lives a lot easier if we do not have to keep track of generated mock types

Reply:
The types defined could just be regular Go structs:
```
type Application struct {
	ID          string
	AppID       string
	Description string
	DisplayName string
}

func (c *AppClient) GetApplication(ctx context.Context, clientID string) (Application, error) {
	filter := fmt.Sprintf("appId eq '%s'", clientID)
	req := applications.ApplicationsRequestBuilderGetRequestConfiguration{
		QueryParameters: &applications.ApplicationsRequestBuilderGetQueryParameters{
			Filter: &filter,
		},
	}

	resp, err := c.client.Applications().Get(ctx, &req)
	if err != nil {
		return Application{}, err
	}

	apps := resp.GetValue()
	if len(apps) == 0 {
		return Application{}, fmt.Errorf("no application found")
	}
	if len(apps) > 1 {
		return Application{}, fmt.Errorf("multiple applications found - double check your client_id")
	}
	app := apps[0]
	application := Application{
		ID:          *app.GetId(),
		AppID:       *app.GetAppId(),
		Description: *app.GetDescription(),
		DisplayName: *app.GetDisplayName(),
	}

	return application, nil
}
```

Reply:
Ah I see what you mean! I think this might work, I might have overcomplicated things in my head initially. Going to give this refactoring a try, might not end up needing the mocks this way 🙏🏼

Reply:
@raymonstah thanks for the extra set of eyes! The refactoring definitely made things simpler and removed the need for mocks altogether 🎉